aio_pika/message.py

```diff
@@ -1,213 +1,221 @@
     return datetime.utcnow() + value
 
 
-@encode_timestamp.register(type(None))
-def _(_):
+@encode_timestamp.register(NoneType)        # type: ignore
+def encode_timestamp_none(_: Any) -> None:
     return None
 
 
 @singledispatch
-def decode_timestamp(value) -> Optional[datetime]:
+def decode_timestamp(value: Any) -> Optional[datetime]:
     raise ValueError("Invalid timestamp type: %r" % type(value), value)
 
 
 @decode_timestamp.register(datetime)
-def _(value):
+def decode_timestamp_datetime(value: datetime) -> datetime:
     return value
 
 
 @decode_timestamp.register(float)
 @decode_timestamp.register(int)
-def _(value):
+def decode_timestamp_number(value: Union[float, int]) -> datetime:
     return datetime.utcfromtimestamp(value)
 
 
 @decode_timestamp.register(time.struct_time)
-def _(value: time.struct_time):
+def decode_timestamp_struct_time(value: time.struct_time) -> datetime:
     return datetime(*value[:6])
 
 
-@decode_timestamp.register(type(None))
-def _(_):
+@decode_timestamp.register(NoneType)    # type: ignore
+def decode_timestamp_none(_: Any) -> None:
     return None
 
 
-def optional(value, func: Callable[[Any], Any] = str, default=None):
-    return func(value) if value else default
+V = TypeVar("V")
+D = TypeVar("D")
+T = TypeVar("T")
 
 
-class HeaderProxy(Mapping):
-    def __init__(self, headers: Dict[str, bytes]):
-        self._headers = headers  # type: Dict[str, bytes]
-        self._cache = {}  # type: Dict[str, Any]
+def optional(
+    value: V,
+    func: Union[Callable[[V], T], Type[T]],
+    default: D = None,
+) -> Union[T, D]:
+    return func(value) if value else default    # type: ignore
 
-    def __getitem__(self, k):
+
+class HeaderProxy(MutableMapping):
+    def __init__(self, headers: FieldTable):
+        self._headers: FieldTable = headers
+        self._cache: Dict[str, Any] = {}
+
+    def __getitem__(self, k: str) -> FieldValue:
         if k not in self._headers:
             raise KeyError(k)
 
         if k not in self._cache:
             value = self._headers[k]
+
             if isinstance(value, bytes):
                 self._cache[k] = value.decode()
             else:
                 self._cache[k] = value
 
         return self._cache[k]
 
-    def __setitem__(self, key, value):
+    def __delitem__(self, key: str) -> None:
+        del self._headers[key]
+
+    def __setitem__(self, key: str, value: FieldValue) -> None:
         self._headers[key] = header_converter(value)
         self._cache.pop(key, None)
 
     def __len__(self) -> int:
         return len(self._headers)
 
-    def __iter__(self):
+    def __iter__(self) -> Iterator[str]:
         for key in self._headers:
             yield key
 
 
 @singledispatch
-def header_converter(value: Any) -> bytes:
-    return json.dumps(
-        value, separators=(",", ":"), ensure_ascii=False, default=repr,
-    ).encode()
+def header_converter(value: Any) -> FieldValue:
+    return bytearray(
+        json.dumps(
```

Comment:
Maybe give options to choose json encoder for client?

Reply:
I don't think that transferring a small number of bytes to the header will not cause a performance problem (or gain).